server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java

```diff
@@ -855,7 +856,10 @@ public DiscoveryNodes build() {
             } else if (this.nodeLeftGeneration != null) {
                 // only happens during deserialization
                 assert removedNode == false;
+                assert resetNodeLeftGeneration == false;
                 newNodeLeftGeneration = nodeLeftGeneration;
+            } else if (resetNodeLeftGeneration) {
+                newNodeLeftGeneration = 0L;
```

Comment:
Should we also set `resetNodeLeftGeneration` to false? Or it does not matter?

Reply:
We basically never re-use builders after calling `.build()` so it won't really matter, but if we did re-use the builder I think we would want to preserve the `resetNodeLeftGeneration` flag.